# Análisis Exploratorio: Hechos de Tránsito (2018-2024)

Este notebook forma parte del pipeline de ciencia de datos del proyecto **crash-severity-predictor**.  
El objetivo es explorar los microdatos oficiales de hechos de tránsito proporcionados por el  
Instituto Nacional de Estadística (INE) para el septenio 2018-2024, identificando la estructura,  
calidad y distribución de las variables antes de cualquier transformación.

> **Fuente:** INE — Microdatos de Hechos de Tránsito (2018-2024)  

In [1]:
# --- Importaciones ---------------------------------------------
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
import warnings

warnings.filterwarnings('ignore')

# Configuración de visualización
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)
sns.set_theme(style='whitegrid', palette='muted')
plt.rcParams['figure.figsize'] = (12, 5)

print('✓ Librerías cargadas correctamente')

✓ Librerías cargadas correctamente


#### 1. Carga de Datos

Se cargan los 7 archivos xlsx correspondientes al septenio 2018-2024.  
Cada archivo representa un año de registros de hechos de tránsito.

In [2]:
# -- Carga de datos ---------------------------------------------
RAW_PATH = '../data/raw/ACCIDENTES DE TRANSITO - HECHOS'
YEARS = range(2018, 2025)

frames = []
for year in YEARS:
    path = os.path.join(RAW_PATH, f'hechos-de-transito-ano-{year}.xlsx')
    df_year = pd.read_excel(path)
    df_year['año_carga'] = year
    frames.append(df_year)
    print(f'  {year}: {len(df_year):,} registros cargados')

df = pd.concat(frames, ignore_index=True)
print(f'\n✓ Total registros: {len(df):,}')

  2018: 6,395 registros cargados
  2019: 7,047 registros cargados
  2020: 6,350 registros cargados
  2021: 8,153 registros cargados
  2022: 7,924 registros cargados
  2023: 8,218 registros cargados
  2024: 8,401 registros cargados

✓ Total registros: 52,488


#### 2. Vista General del Dataset

In [3]:
# -- Dimensiones y tipos de dato ---------------------------------------------
print(f'Dimensiones: {df.shape[0]:,} filas × {df.shape[1]} columnas')
print(f'\nColumnas y tipos de dato:')
print(df.dtypes)
print(f'\nPrimeras 3 filas:')
df.head(3)

Dimensiones: 52,488 filas × 27 columnas

Columnas y tipos de dato:
núm_corre       float64
año_ocu         float64
día_ocu         float64
hora_ocu        float64
g_hora            int64
g_hora_5          int64
mes_ocu         float64
día_sem_ocu     float64
mupio_ocu         int64
depto_ocu         int64
zona_ocu          int64
tipo_veh          int64
marca_veh         int64
color_veh         int64
modelo_veh        int64
g_modelo_veh      int64
tipo_eve          int64
año_carga         int64
Núm_corre       float64
Año_ocu         float64
Día_ocu         float64
Hora_ocu        float64
Mes_ocu         float64
zona_ciudad     float64
num_corre       float64
dia_ocu         float64
dia_sem_ocu     float64
dtype: object

Primeras 3 filas:


,núm_corre,año_ocu,día_ocu,hora_ocu,g_hora,g_hora_5,mes_ocu,día_sem_ocu,mupio_ocu,depto_ocu,zona_ocu,tipo_veh,marca_veh,color_veh,modelo_veh,g_modelo_veh,tipo_eve,año_carga,Núm_corre,Año_ocu,Día_ocu,Hora_ocu,Mes_ocu,zona_ciudad,num_corre,dia_ocu,dia_sem_ocu
0,1.00,2018.00,1.00,16.00,3,2,1.00,1.00,115,1,99,4,32,5,9999,99,2,2018,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2.00,2018.00,1.00,12.00,3,2,1.00,1.00,2207,22,99,1,69,2,9999,99,1,2018,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,3.00,2018.00,1.00,7.00,2,1,1.00,1.00,2102,21,99,1,999,6,9999,99,2,2018,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [4]:
# -- Columnas por año ---------------------------------------------
for year in YEARS:
    path = os.path.join(RAW_PATH, f'hechos-de-transito-ano-{year}.xlsx')
    cols = pd.read_excel(path, nrows=0).columns.tolist()
    print(f'\n{year}: {cols}')


2018: ['núm_corre', 'año_ocu', 'día_ocu', 'hora_ocu', 'g_hora', 'g_hora_5', 'mes_ocu', 'día_sem_ocu', 'mupio_ocu', 'depto_ocu', 'zona_ocu', 'tipo_veh', 'marca_veh', 'color_veh', 'modelo_veh', 'g_modelo_veh', 'tipo_eve']

2019: ['núm_corre', 'año_ocu', 'día_ocu', 'hora_ocu', 'g_hora', 'g_hora_5', 'mes_ocu', 'día_sem_ocu', 'depto_ocu', 'mupio_ocu', 'zona_ocu', 'tipo_veh', 'marca_veh', 'color_veh', 'modelo_veh', 'g_modelo_veh', 'tipo_eve']

2020: ['Núm_corre', 'Año_ocu', 'Día_ocu', 'Hora_ocu', 'g_hora', 'g_hora_5', 'Mes_ocu', 'día_sem_ocu', 'depto_ocu', 'mupio_ocu', 'zona_ocu', 'tipo_veh', 'marca_veh', 'color_veh', 'modelo_veh', 'g_modelo_veh', 'tipo_eve']

2021: ['núm_corre', 'año_ocu', 'día_ocu', 'hora_ocu', 'g_hora', 'g_hora_5', 'mes_ocu', 'día_sem_ocu', 'depto_ocu', 'mupio_ocu', 'zona_ocu', 'zona_ciudad', 'tipo_veh', 'marca_veh', 'color_veh', 'modelo_veh', 'g_modelo_veh', 'tipo_eve']

2022: ['num_corre', 'año_ocu', 'día_ocu', 'hora_ocu', 'g_hora', 'g_hora_5', 'mes_ocu', 'día_sem_ocu'

### Hallazgo 1 : Inconsistencia en nombres de columnas

El INE cambió los nombres de columnas entre años:

| Columna estándar | Variantes encontradas | Años |
|---|---|---|
| `num_corre` | `núm_corre`, `Núm_corre`, `num_corre` | 2018-2019, 2020, 2022-2024 |
| `dia_ocu` | `día_ocu`, `Día_ocu`, `dia_ocu` | 2018-2022, 2020, 2023-2024 |
| `dia_sem_ocu` | `día_sem_ocu`, `dia_sem_ocu` | 2018-2022, 2023-2024 |
| `mes_ocu` | `mes_ocu`, `Mes_ooo` | 2018-2019/2021-2024, 2020 |
| `hora_ocu` | `hora_ocu`, `Hora_ocu` | 2018-2019/2021-2024, 2020 |

**Columnas extra:**
- `zona_ciudad` aparece únicamente en 2021 — requiere investigación
- 2020 tiene todas las columnas identificadoras en mayúscula

**Acción en limpieza:** estandarizar todos los nombres a minúscula sin tildes.

In [5]:
# -- Validar si zona_ciudad == zona_ocu en 2021 ---------------------------------------------
df_2021 = pd.read_excel(
    os.path.join(RAW_PATH, 'hechos-de-transito-ano-2021.xlsx')
)

iguales = (df_2021['zona_ocu'] == df_2021['zona_ciudad']).all()
print(f'¿zona_ocu == zona_ciudad en todos los registros? {iguales}')
print(f'\nzona_ocu únicos:     {sorted(df_2021["zona_ocu"].unique())}')
print(f'zona_ciudad únicos:  {sorted(df_2021["zona_ciudad"].unique())}')

¿zona_ocu == zona_ciudad en todos los registros? False

zona_ocu únicos:     [np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9), np.int64(10), np.int64(11), np.int64(12), np.int64(13), np.int64(14), np.int64(15), np.int64(16), np.int64(17), np.int64(18), np.int64(19), np.int64(21), np.int64(24), np.int64(25), np.int64(99)]
zona_ciudad únicos:  [np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9), np.int64(10), np.int64(11), np.int64(12), np.int64(13), np.int64(14), np.int64(15), np.int64(16), np.int64(17), np.int64(18), np.int64(19), np.int64(21), np.int64(24), np.int64(25), np.int64(99)]


In [6]:
# -- Ver registros donde difieren ---------------------------------------------
difieren = df_2021[df_2021['zona_ocu'] != df_2021['zona_ciudad']]
print(f'Registros donde difieren: {len(difieren):,} de {len(df_2021):,}')
print(f'\nEjemplos:')
print(difieren[['núm_corre', 'zona_ocu', 'zona_ciudad']].head(10))

Registros donde difieren: 1,342 de 8,153

Ejemplos:
    núm_corre  zona_ocu  zona_ciudad
6           7         4           99
7           8         4           99
10         11         6           99
19         20         4           99
29         30         2           99
34         35         4           99
35         36         5           99
47         48         5           99
54         55         3           99
66         67        12           99


### Hallazgo 2 : zona_ciudad vs zona_ocu (2021)

`zona_ciudad` y `zona_ocu` **no son la misma variable**.  
En 1,342 de 8,153 registros (16.5%) los valores difieren.  
Cuando `zona_ocu` tiene valor real, `zona_ciudad` registra 99 (ignorado) y viceversa.

**Conclusión:** son dos campos complementarios del mismo concepto geográfico.  
En la limpieza se consolidarán en una sola columna `zona_ocu` tomando  
el valor real de cualquiera de las dos cuando el otro sea 99.

In [7]:
# -- Confirmar que son complementarias ---------------------------------------------
ambas_99 = df_2021[
    (df_2021['zona_ocu'] == 99) & (df_2021['zona_ciudad'] == 99)
]
ambas_valor = df_2021[
    (df_2021['zona_ocu'] != 99) & (df_2021['zona_ciudad'] != 99)
]
print(f'Ambas en 99 (ignorado):      {len(ambas_99):,}')
print(f'Ambas con valor real:         {len(ambas_valor):,}')
print(f'Solo zona_ocu tiene valor:    {len(df_2021[(df_2021["zona_ocu"] != 99) & (df_2021["zona_ciudad"] == 99)]):,}')
print(f'Solo zona_ciudad tiene valor: {len(df_2021[(df_2021["zona_ocu"] == 99) & (df_2021["zona_ciudad"] != 99)]):,}')

Ambas en 99 (ignorado):      5,112
Ambas con valor real:         1,699
Solo zona_ocu tiene valor:    1,342
Solo zona_ciudad tiene valor: 0


### Conclusión zona_ciudad (2021)

Los resultados confirman que `zona_ciudad` es completamente redundante:
- Nunca aporta información que `zona_ocu` no tenga (0 registros exclusivos)
- En 1,342 registros `zona_ocu` tiene valor mientras `zona_ciudad` es 99

**Acción en limpieza:** eliminar `zona_ciudad` del año 2021.  
`zona_ocu` es suficiente para el septenio completo.

## 3. Valores Ignorados y Nulos

In [8]:
# -- Valores nulos reales ----------------------------------------------
print('Valores nulos por columna:')
print(df.isnull().sum().sort_values(ascending=False))

Valores nulos por columna:
Día_ocu         46138
Hora_ocu        46138
Núm_corre       46138
Año_ocu         46138
Mes_ocu         46138
zona_ciudad     44335
dia_ocu         35869
dia_sem_ocu     35869
núm_corre       30893
num_corre       27945
día_ocu         22969
día_sem_ocu     16619
mes_ocu          6350
hora_ocu         6350
año_ocu          6350
g_hora              0
depto_ocu           0
mupio_ocu           0
g_hora_5            0
zona_ocu            0
año_carga           0
g_modelo_veh        0
tipo_eve            0
marca_veh           0
color_veh           0
tipo_veh            0
modelo_veh          0
dtype: int64


### Hallazgo 3 : Nulos aparentes por inconsistencia de nombres

Los nulos no representan datos faltantes reales sino columnas duplicadas
por cambios de nomenclatura entre años:

| Columna | Nulos | Causa |
|---|---|---|
| `Núm_corre`, `núm_corre`, `num_corre` | 46,138 / 30,893 / 27,945 | Mismo campo, 3 nombres distintos |
| `Día_ocu`, `día_ocu`, `dia_ocu` | 46,138 / 22,969 / 35,869 | Mismo campo, mayúsculas y tildes |
| `Mes_ocu`, `mes_ocu` | 46,138 / 6,350 | 2020 en mayúscula |
| `zona_ciudad` | 44,335 | Solo existe en 2021 |

**Conclusión:** no hay nulos reales en columnas críticas.  
El problema es de estandarización de nombres, no de calidad de datos.  
**Acción en limpieza:** normalizar todos los nombres a minúscula sin tildes.

In [9]:
# -- Verificar ignorados por variable crítica ----------------------------------------------
IGNORE_CODES = {
    'tipo_veh'    : 99,
    'g_hora_5'    : 4,
    'g_hora'      : 5,
    'zona_ocu'    : 99,
    'tipo_eve'    : 99,
    'marca_veh'   : 999,
    'color_veh'   : 99,
    'modelo_veh'  : 9999,
    'g_modelo_veh': 99,
}

print(f'{"Variable":<15} {"Ignorados":>10} {"% del total":>12}')
print('-' * 40)
for col, code in IGNORE_CODES.items():
    if col in df.columns:
        n = (df[col] == code).sum()
        pct = n / len(df) * 100
        print(f'{col:<15} {n:>10,} {pct:>11.1f}%')

Variable         Ignorados  % del total
----------------------------------------
tipo_veh             2,691         5.1%
g_hora_5                29         0.1%
g_hora                  29         0.1%
zona_ocu            36,201        69.0%
tipo_eve                 9         0.0%
marca_veh           16,122        30.7%
color_veh           14,791        28.2%
modelo_veh          33,702        64.2%
g_modelo_veh        33,703        64.2%


### Hallazgo 4 : Análisis de códigos ignorados por variable

| Variable | Ignorados | % | Decisión en limpieza |
|---|---|---|---|
| `tipo_veh` | 2,691 | 5.1% | Imputar con moda |
| `g_hora_5` | 29 | 0.1% | Imputar con moda |
| `g_hora` | 29 | 0.1% | Imputar con moda |
| `tipo_eve` | 9 | 0.0% | Imputar con moda |
| `marca_veh` | 16,122 | 30.7% | Mantener como categoría "Ignorado" |
| `color_veh` | 14,791 | 28.2% | Mantener como categoría "Ignorado" |
| `zona_ocu` | 36,201 | 69.0% | Mantener como categoría "Ignorado" |
| `modelo_veh` | 33,702 | 64.2% | Mantener como categoría "Ignorado" |
| `g_modelo_veh` | 33,703 | 64.2% | Mantener como categoría "Ignorado" |

**Observaciones críticas:**
- `zona_ocu` tiene 69% ignorados — no es una variable confiable para el modelo
- `modelo_veh` y `g_modelo_veh` tienen 64% ignorados — mismo caso
- `marca_veh` y `color_veh` con 30% — se conservan como categoría pero no serán variables predictoras principales
- `tipo_veh`, `g_hora_5` y `tipo_eve` tienen menos del 6% — variables confiables y críticas para el modelo

## 4. Distribución de Variables

### 4.1 Distribución de tipo_veh

In [14]:
# -- Distribución de tipo_eve ----------------------------------------------
print(df['tipo_eve'].value_counts().sort_index())

tipo_eve
1     27831
2      3563
3      1135
4       331
5     14167
6      5058
7       367
8        27
99        9
Name: count, dtype: int64


### Distribución de tipo_eve (Tipo de Evento)

| Código | Tipo | Registros | % |
|---|---|---|---|
| 1 | Colisión | 27,831 | 53.0% |
| 2 | Choque | 3,563 | 6.8% |
| 3 | Vuelco | 1,135 | 2.2% |
| 4 | Caída | 331 | 0.6% |
| 5 | Atropello | 14,167 | 27.0% |
| 6 | Derrape | 5,058 | 9.6% |
| 7 | Embarranco | 367 | 0.7% |
| 8 | Encuneto | 27 | 0.1% |
| 99 | Ignorado | 9 | 0.0% |

Colisión y Atropello concentran el **80%** de los hechos.

In [15]:
# -- Distribución de tipo_veh ----------------------------------------------
print(df['tipo_veh'].value_counts().sort_index())

tipo_veh
1     12272
2      4359
3      5925
4     18412
5      3124
6      1966
7       191
8        53
9       996
10       24
11      296
12     1475
13       36
14      456
16       12
18       26
20        1
22      168
23        4
24        1
99     2691
Name: count, dtype: int64


### Distribución de tipo_veh (Tipo de Vehículo)

| Código | Tipo | Registros | % |
|---|---|---|---|
| 1 | Automóvil | 12,272 | 23.4% |
| 2 | Camioneta sport/blazer | 4,359 | 8.3% |
| 3 | Pick up | 5,925 | 11.3% |
| 4 | Motocicleta | 18,412 | 35.1% |
| 5 | Camión | 3,124 | 5.9% |
| 6 | Cabezal | 1,966 | 3.7% |
| 7 | Bus extraurbano | 191 | 0.4% |
| 8 | Jeep | 53 | 0.1% |
| 9 | Microbús | 996 | 1.9% |
| 12 | Bus urbano | 1,475 | 2.8% |
| 14 | Mototaxi | 456 | 0.9% |
| 99 | Ignorado | 2,691 | 5.1% |
| Resto | (11 tipos minoritarios) | 303 | 0.6% |

**Motocicleta domina con 35.1%** — variable crítica para el modelo.  
Automóvil + Pick up + Motocicleta concentran el **70%**.

In [16]:
# -- Distribución de g_hora_5 ----------------------------------------------
print(df['g_hora_5'].value_counts().sort_index())

g_hora_5
1    18651
2    13860
3    19948
4       29
Name: count, dtype: int64


### Distribución de g_hora_5 (Grupo Horario)

| Código | Grupo | Registros | % |
|---|---|---|---|
| 1 | Mañana | 18,651 | 35.5% |
| 2 | Tarde | 13,860 | 26.4% |
| 3 | Noche | 19,948 | 38.0% |
| 4 | Ignorado | 29 | 0.1% |

**Noche y Mañana concentran el 73.5%** de los hechos.  
Variable muy confiable — solo 29 ignorados (0.1%).

In [17]:
# -- Distribución anual de hechos ----------------------------------------------
print(df['año_carga'].value_counts().sort_index())

año_carga
2018    6395
2019    7047
2020    6350
2021    8153
2022    7924
2023    8218
2024    8401
Name: count, dtype: int64


### Distribución anual de hechos (2018-2024)

| Año | Registros | % |
|---|---|---|
| 2018 | 6,395 | 12.2% |
| 2019 | 7,047 | 13.4% |
| 2020 | 6,350 | 12.1% |
| 2021 | 8,153 | 15.5% |
| 2022 | 7,924 | 15.1% |
| 2023 | 8,218 | 15.6% |
| 2024 | 8,401 | 16.0% |

**Total: 52,488 registros.**  
2020 muestra la menor frecuencia — posible efecto pandemia.  
Tendencia creciente sostenida desde 2021.

In [18]:
# -- Distribución por dia_sem_ocu ----------------------------------------------
# Normalizar primero la columna fragmentada
dia_cols = [c for c in df.columns if 'dia_sem' in c.lower() or 'día_sem' in c.lower()]
print('Columnas encontradas:', dia_cols)

df['dia_sem_norm'] = df[dia_cols].bfill(axis=1).iloc[:, 0]
print(df['dia_sem_norm'].value_counts().sort_index())

Columnas encontradas: ['día_sem_ocu', 'dia_sem_ocu']
dia_sem_norm
1.00     6957
2.00     6008
3.00     5976
4.00     6326
5.00     7397
6.00    10080
7.00     9744
Name: count, dtype: int64


### Distribución por día de la semana

| Código | Día | Registros | % |
|---|---|---|---|
| 1 | Lunes | 6,957 | 13.3% |
| 2 | Martes | 6,008 | 11.4% |
| 3 | Miércoles | 5,976 | 11.4% |
| 4 | Jueves | 6,326 | 12.1% |
| 5 | Viernes | 7,397 | 14.1% |
| 6 | Sábado | 10,080 | 19.2% |
| 7 | Domingo | 9,744 | 18.6% |

**Sábado y Domingo concentran el 37.8%** de los hechos.  
Viernes + fin de semana suman el **51.9%** — patrón de riesgo claro.  
La columna quedó correctamente consolidada desde las 2 variantes.

In [19]:
# -- Top 10 municipios con más hechos ----------------------------------------------
print(df['mupio_ocu'].value_counts().head(10))

mupio_ocu
101     11362
108      2726
115      2385
501      1841
116      1001
114       836
117       824
401       771
1601      771
110       730
Name: count, dtype: int64


### Top 10 municipios con más hechos

| Código | Municipio | Depto | Registros | % |
|---|---|---|---|---|
| 101 | Guatemala | Guatemala | 11,362 | 21.6% |
| 108 | Mixco | Guatemala | 2,726 | 5.2% |
| 115 | Villa Nueva | Guatemala | 2,385 | 4.5% |
| 501 | Escuintla | Escuintla | 1,841 | 3.5% |
| 116 | Villa Canales | Guatemala | 1,001 | 1.9% |
| 114 | Amatitlán | Guatemala | 836 | 1.6% |
| 117 | San Miguel Petapa | Guatemala | 824 | 1.6% |
| 401 | Chimaltenango | Chimaltenango | 771 | 1.5% |
| 1601 | Cobán | Alta Verapaz | 771 | 1.5% |
| 110 | San Juan Sacatepéquez | Guatemala | 730 | 1.4% |

**Guatemala capital concentra el 21.6%** de todos los hechos.  
7 de los 10 municipios críticos pertenecen al depto de Guatemala.

In [20]:
# -- resumen de variables críticas ----------------------------------------------
vars_criticas = {
    'tipo_eve'   : 'Confiable — 0.0% ignorados',
    'g_hora_5'   : 'Confiable — 0.1% ignorados',
    'dia_sem_norm': 'Confiable — 0 nulos tras consolidación',
    'tipo_veh'   : 'Confiable — 5.1% ignorados',
    'mupio_ocu'  : 'Confiable — 0 nulos',
    'marca_veh'  : 'Limitada — 30.7% ignorados',
    'color_veh'  : 'Limitada — 28.2% ignorados',
    'zona_ocu'   : 'No confiable — 69.0% ignorados',
    'modelo_veh' : 'No confiable — 64.2% ignorados',
}

print(f'{"Variable":<20} {"Calidad"}')
print('-' * 50)
for k, v in vars_criticas.items():
    print(f'{k:<20} {v}')

Variable             Calidad
--------------------------------------------------
tipo_eve             Confiable — 0.0% ignorados
g_hora_5             Confiable — 0.1% ignorados
dia_sem_norm         Confiable — 0 nulos tras consolidación
tipo_veh             Confiable — 5.1% ignorados
mupio_ocu            Confiable — 0 nulos
marca_veh            Limitada — 30.7% ignorados
color_veh            Limitada — 28.2% ignorados
zona_ocu             No confiable — 69.0% ignorados
modelo_veh           No confiable — 64.2% ignorados


## 5. Relaciones entre Variables Críticas

In [21]:
# -- tipo_eve vs g_hora_5 ----------------------------------------------
cross = pd.crosstab(
    df['tipo_eve'],
    df['g_hora_5'],
    normalize='index'
).round(3) * 100

cross.columns = ['Mañana', 'Tarde', 'Noche', 'Ignorado']
cross.index = ['Colisión','Choque','Vuelco','Caída',
               'Atropello','Derrape','Embarranco','Encuneto','Ignorado']
print(cross.drop(columns='Ignorado').to_string())

            Mañana  Tarde  Noche
Colisión     34.70  28.10  37.10
Choque       44.70  21.80  33.50
Vuelco       39.90  33.50  26.60
Caída        36.60  26.30  37.20
Atropello    31.50  26.00  42.40
Derrape      43.40  19.60  36.90
Embarranco   41.40  31.90  26.70
Encuneto     22.20  37.00  40.70
Ignorado     55.60   0.00  44.40


### tipo_eve vs g_hora_5

| Tipo | Mañana | Tarde | Noche |
|---|---|---|---|
| Colisión | 34.7% | 28.1% | 37.1% |
| Choque | 44.7% | 21.8% | 33.5% |
| Vuelco | 39.9% | 33.5% | 26.6% |
| Caída | 36.6% | 26.3% | 37.2% |
| Atropello | 31.5% | 26.0% | **42.4%** |
| Derrape | 43.4% | 19.6% | 36.9% |
| Embarranco | 41.4% | 31.9% | 26.7% |

**Patrones relevantes:**
- Atropello se concentra en noche (42.4%) — el más nocturno de todos
- Choque y Derrape son predominantemente matutinos (~43%)
- Vuelco y Embarranco son los más diurnos — probable relación con carretera

In [23]:
# -- tipo_eve vs tipo_veh (top 5 vehículos) ----------------------------------------------
top_veh = [1, 2, 3, 4, 5]
labels = {1:'Automóvil', 2:'Camioneta', 3:'Pick up', 4:'Moto', 5:'Camión'}

cross2 = pd.crosstab(
    df[df['tipo_veh'].isin(top_veh) & (df['tipo_eve'] != 99)]['tipo_eve'],
    df[df['tipo_veh'].isin(top_veh) & (df['tipo_eve'] != 99)]['tipo_veh'],
    normalize='columns'
).round(3) * 100

cross2.columns = [labels[c] for c in cross2.columns]
cross2.index = ['Colisión','Choque','Vuelco','Caída',
                'Atropello','Derrape','Embarranco','Encuneto']
print(cross2.to_string())

            Automóvil  Camioneta  Pick up  Moto  Camión
Colisión        58.90      71.10    61.90 40.90   65.30
Choque           6.70       6.50     8.20  7.70    6.10
Vuelco           1.70       2.50     5.70  0.10    4.60
Caída            0.10       0.30     0.40  1.40    0.50
Atropello       31.70      18.20    21.00 23.50   21.50
Derrape          0.50       0.40     0.90 26.10    0.70
Embarranco       0.40       0.90     1.90  0.20    1.20
Encuneto         0.10       0.10     0.10  0.00    0.10


### tipo_eve vs tipo_veh (top 5 vehículos)

| Tipo | Automóvil | Camioneta | Pick up | Moto | Camión |
|---|---|---|---|---|---|
| Colisión | 58.9% | 71.1% | 61.9% | 40.9% | 65.3% |
| Choque | 6.7% | 6.5% | 8.2% | 7.7% | 6.1% |
| Vuelco | 1.7% | 2.5% | 5.7% | 0.1% | 4.6% |
| Caída | 0.1% | 0.3% | 0.4% | **1.4%** | 0.5% |
| Atropello | 31.7% | 18.2% | 21.0% | 23.5% | 21.5% |
| Derrape | 0.5% | 0.4% | 0.9% | **26.1%** | 0.7% |
| Embarranco | 0.4% | 0.9% | 1.9% | 0.2% | 1.2% |

**Patrones relevantes:**
- Moto tiene un perfil completamente distinto al resto — Derrape (26.1%) es casi exclusivo de motos
- Camioneta es la más propensa a Colisión (71.1%)
- Pick up y Camión lideran en Vuelco (~5%) — masa y velocidad en carretera
- Atropello es relevante en todos los tipos, especialmente Automóvil (31.7%)

In [24]:
# -- dia_sem_norm vs tipo_eve ----------------------------------------------
cross3 = pd.crosstab(
    df[df['tipo_eve'] != 99]['dia_sem_norm'],
    df[df['tipo_eve'] != 99]['tipo_eve'],
    normalize='index'
).round(3) * 100

cross3.columns = ['Colisión','Choque','Vuelco','Caída',
                  'Atropello','Derrape','Embarranco','Encuneto']
cross3.index = ['Lunes','Martes','Miércoles','Jueves',
                'Viernes','Sábado','Domingo']
print(cross3.to_string())

           Colisión  Choque  Vuelco  Caída  Atropello  Derrape  Embarranco  Encuneto
Lunes         51.60    6.40    2.30   0.60      28.50     9.80        0.70      0.00
Martes        55.90    5.60    2.30   0.80      27.10     7.40        0.90      0.00
Miércoles     55.90    6.50    2.10   0.50      26.60     7.70        0.60      0.10
Jueves        54.20    6.70    2.40   0.60      27.70     7.90        0.60      0.00
Viernes       54.20    6.00    2.00   0.60      27.40     9.00        0.80      0.10
Sábado        52.60    6.70    1.90   0.50      26.50    11.10        0.60      0.10
Domingo       49.40    8.70    2.20   0.90      25.80    12.20        0.70      0.10


### dia_sem vs tipo_eve

| Día | Colisión | Atropello | Derrape | Choque |
|---|---|---|---|---|
| Lunes | 51.6% | 28.5% | 9.8% | 6.4% |
| Martes | 55.9% | 27.1% | 7.4% | 5.6% |
| Miércoles | 55.9% | 26.6% | 7.7% | 6.5% |
| Jueves | 54.2% | 27.7% | 7.9% | 6.7% |
| Viernes | 54.2% | 27.4% | 9.0% | 6.0% |
| Sábado | 52.6% | 26.5% | **11.1%** | 6.7% |
| Domingo | 49.4% | 25.8% | **12.2%** | **8.7%** |

**Patrones relevantes:**
- Derrape escala progresivamente hacia el fin de semana — pico domingo (12.2%)
- Choque también sube el domingo (8.7%) — probable conducción bajo efectos de alcohol
- Colisión es más alta entre semana (55.9% martes/miércoles) — tráfico laboral

In [25]:
# -- Cierre sección 5: correlación entre variables numéricas ----------------------------------------------
num_cols = ['tipo_eve', 'tipo_veh', 'g_hora_5', 'dia_sem_norm', 'mupio_ocu']
print(df[num_cols].corr().round(3))

              tipo_eve  tipo_veh  g_hora_5  dia_sem_norm  mupio_ocu
tipo_eve          1.00      0.14      0.01          0.02      -0.01
tipo_veh          0.14      1.00      0.00          0.00       0.00
g_hora_5          0.01      0.00      1.00          0.01       0.05
dia_sem_norm      0.02      0.00      0.01          1.00       0.00
mupio_ocu        -0.01      0.00      0.05          0.00       1.00


### Correlación entre variables críticas

Correlaciones lineales bajas entre todas las variables — esperado,
ya que son variables categóricas codificadas numéricamente.
La relación `tipo_eve` ↔ `tipo_veh` (0.14) es la única señal,
consistente con el crosstab anterior (Moto → Derrape).

**Conclusión:** las relaciones entre variables son no lineales.
El modelo deberá capturar interacciones — árbol de decisión o
random forest más adecuados que regresión lineal.

## 6. Resumen EDA

In [26]:
# -- Resumen final ----------------------------------------------
resumen = {
    'Total registros'            : f'{len(df):,}',
    'Período'                    : '2018–2024',
    'Variables confiables'       : 'tipo_eve, g_hora_5, dia_sem, tipo_veh, mupio_ocu',
    'Variables limitadas'        : 'marca_veh (30.7%), color_veh (28.2%)',
    'Variables descartadas'      : 'zona_ocu (69%), modelo_veh (64%)',
    'Problema principal'         : 'Inconsistencia de nombres entre años',
    'Acción limpieza pendiente'  : 'Normalizar columnas + eliminar zona_ciudad',
}

print(f'{"Campo":<30} {"Valor"}')
print('-' * 65)
for k, v in resumen.items():
    print(f'{k:<30} {v}')

Campo                          Valor
-----------------------------------------------------------------
Total registros                52,488
Período                        2018–2024
Variables confiables           tipo_eve, g_hora_5, dia_sem, tipo_veh, mupio_ocu
Variables limitadas            marca_veh (30.7%), color_veh (28.2%)
Variables descartadas          zona_ocu (69%), modelo_veh (64%)
Problema principal             Inconsistencia de nombres entre años
Acción limpieza pendiente      Normalizar columnas + eliminar zona_ciudad


### Resumen EDA : Hechos de Tránsito 2018–2024

**Dataset:** 52,488 registros × 7 años

**Variables confiables para el modelo:**
- `tipo_eve` — 8 tipos de evento, 0.0% ignorados
- `tipo_veh` — motocicleta domina (35.1%), 5.1% ignorados
- `g_hora_5` — noche y mañana concentran 73.5%, 0.1% ignorados
- `dia_sem_norm` — fin de semana concentra 37.8%, 0 nulos
- `mupio_ocu` — capital concentra 21.6%, 0 nulos

**Variables limitadas (conservar como categoría):**
- `marca_veh` — 30.7% ignorados
- `color_veh` — 28.2% ignorados

**Variables descartadas del modelo:**
- `zona_ocu` — 69.0% ignorados
- `modelo_veh` / `g_modelo_veh` — 64.2% ignorados

**Problemas de calidad a resolver en ETL:**
1. Normalizar nombres de columnas (tildes, mayúsculas, variantes)
2. Eliminar `zona_ciudad` del año 2021 (redundante)

**Hallazgos clave:**
- Moto tiene perfil de riesgo único — Derrape exclusivo (26.1%)
- Atropello se concentra en noche (42.4%)
- Fin de semana eleva Derrape y Choque — posible alcohol
- Relaciones no lineales → modelo basado en árboles recomendado